### DISCUSSION TITLE
DecoderSID?

### DISCUSSION THREAD

**Discussion by beachviking** [pschatzmann/arduino-audiokit#72] [Ideas]
Hello @pschatzmann,

Your libraries came to my attention through Andreas Spiess' videos and I must say I have some fun playing with your libraries!

I also came across this project recently:
https://github.com/possan/csid-esp32

Which made me wanted to see if I was able to make my AI Audio Kit V2.2 use your libraries and be able to play old C64 sid files.

I was somewhat successful.  I took some inspiration by looking at your Sinewave example and was able to make libcsid library kind of work.  Sounds great!  However, the code uses a sid tune that is stored in memory.  I tried to see how I would go about loading up an SD card with a bunch of SID tunes and play these, similarly to the MP3 examples you have provided.  This proved difficult with my limited knowledge of all the libraries about the best way to go about doing this.  I also noted that there seems to be an assumption in the libcsid library that the whole tune must be in memory first before it can play it, as it uses extensive memory copying mechanisms to set up the simulation environment.

Just wanted to reach out to you to see if you have any pointers on how to go about solving this and/or if you want to include a SID decoder in your library. :)

[esp32-memory-sid-audiokit.zip](https://github.com/pschatzmann/arduino-audiokit/files/10854358/esp32-memory-sid-audiokit.zip)

Kind regards from Norway,

Gunnar Larsen

**Comment by pschatzmann** [comment]
Cool, I like the idea. I would propose to use the AudioTools and not the AudioKit as starting point. This has the advantage that it will play not only on an AudioKit but on any other supported device as well and that you can store the data on different media.

As far as the file size is concearned, I would  not be too worried because each file just has a couple of K, so it should be fine to allocate this in RAM (and on the AudioKit we can even use the 4MB PSRAM). I suggest to use a Vector<uint8> where we can just call a resize() with the actual file size to load.

Unfortunately this does not really fit into the current Codec Architecture. So it will be something separate. 
I noticed that it is easy to download hundreds or thousands of small files, so I would suggest to store them on an SD drive or in one of the drive format supported by the AudioPlayer class; This way we can use the existing file navigation logic.

So I propose to create a SIDPlayer class that relies on the existing functionality of the AudioTools AudioPlayer (using the parameters AudioSource and AudioPrint) and just implements the methods which are different:
- custom copy logic
- decoder

ps. here are some hits on how to store the songs in Progmem: https://github.com/pschatzmann/arduino-audio-tools/wiki/MemoryStream:-Converting-a-File-to-Flash-Memory
 
ps. I just noticed that your arrays are missing a const: this is causing that they are stored in RAM instead of PROGMEM

**Reply by beachviking** [reply]
Hello again, thank you for your suggestions, I will definitely have a look at the proposed approach and see if I can make it work.  Thanks! Will share progress as appropriate. :)

**Comment by pschatzmann** [comment]
Actually I started to work on it. Maybe we can do this together ?

**Reply by beachviking** [reply]
Awesome!  Sure, I will help in any way I can!

**Reply by beachviking** [reply]
I did find a potential better starting point for the SID emulation, seems to be cleaned up a bit.  Have not tried it yet.
https://github.com/mlund/cppsid

**Comment by pschatzmann** [comment]
Have a look at https://github.com/pschatzmann/SIDPlayer
Todos:
- extend constructor of SDPlayer with class to determine the size: DONE
- fix compile error
- provide README
- debugging/testing

**Reply by pschatzmann** [reply]
I have a concept now to determine the size: Can you help with the open points above.

**Comment by beachviking** [comment]
Sure thing. Fairly busy during the day, but will have some time tonight.Sent from my iPhoneOn 1 Mar 2023, at 10:07, Phil Schatzmann ***@***.***> wrote:﻿
I have a concept now to determine the size: Can you help with the open points above.

**Reply by pschatzmann** [reply]
No worries...

**Comment by beachviking** [comment]
Hello again, I think I managed to fix the compile issues.  New file attached for the libcsidlight.cpp file.

[src.zip](https://github.com/pschatzmann/arduino-audiokit/files/10864101/src.zip)

The ESP32 does however enter a reset loop with the following messages:
`Rebooting...
ets Jul 29 2019 12:21:46

rst:0xc (SW_CPU_RESET),boot:0x1f (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13104
load:0x40080400,len:3036
entry 0x400805e4
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.

Core  0 register dump:
PC      : 0x400d3417  PS      : 0x00060f30  A0      : 0x800d8444  A1      : 0x3ffe3bb0  
A2      : 0x00000000  A3      : 0x3ffe3be0  A4      : 0x00000001  A5      : 0x00000000
A6      : 0x3ffca894  A7      : 0x3ffca988  A8      : 0x00000000  A9      : 0x3ffe3ba0  
A10     : 0x00000000  A11     : 0x3ffe3be0  A12     : 0x00000001  A13     : 0x00000000
A14     : 0x3ffcb654  A15     : 0x3ffc7688  SAR     : 0x00000011  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000000  LBEG    : 0x400d57f4  LEND    : 0x400d5800  LCOUNT  : 0x00000000


Backtrace: 0x400d3414:0x3ffe3bb0 0x400d8441:0x3ffe3be0 0x400e6603:0x3ffe3c10 0x40082e4e:0x3ffe3c40 0x400792ba:0x3ffe3c90 |<-CORRUPTED

  #0  0x400d3414:0x3ffe3bb0 in audio_tools::SIDPlayer::setAudioInfo(audio_tools::AudioBaseInfo) at src/SIDPlayer.h:46
  #1  0x400d8441:0x3ffe3be0 in audio_tools::SIDPlayer::SIDPlayer(audio_tools::AudioSource&, audio_tools::AudioStream&, audio_tools::SizeSource&) at src/SIDPlayer.h:25
      (inlined by) __static_initialization_and_destruction_0 at src/main.cpp:10
      (inlined by) _GLOBAL__sub_I_Comic_Bakery_sid_len at src/main.cpp:22
  #2  0x400e6603:0x3ffe3c10 in do_global_ctors at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/startup.c:199
      (inlined by) start_cpu0_default at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/startup.c:436
  #3  0x40082e4e:0x3ffe3c40 in call_start_cpu0 at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/port/cpu_start.c:632
  #4  0x400792ba:0x3ffe3c90 in ?? ??:0

`

I also took the liberty to remove files that do not work well due to limitations in the player code.  It does not support emulation of a CIA IC and volume sample tricks used by some songs, which require either a real Commodore 64 or more advanced emulator logic. Ref. https://gist.github.com/cbmeeks/2b107f0a8d36fc461ebb056e94b2f4d6

**Reply by beachviking** [reply]
I also noted in the SID file spec that each song can actually contain multiple tunes. In our code we hardcode the tune number to play. I can add code to the library to expose the number of tunes present in a SID file.

**Reply by pschatzmann** [reply]
Hallo, can you clone my project and contribute with Pull Requests ?
Let's do the discussion about the tune number after we managed to have the first tune working...

**Reply by beachviking** [reply]
Hello, I think I managed to create a pull request on my changes.  I had to fork the project as I do not have write privileges. Hopefully I did things ok. :)

**Reply by pschatzmann** [reply]
Yes, it was working perfectly and as you can see I have merged your changes. You can also use pull requests to get my latest changes.
I also started a new discussions: https://github.com/pschatzmann/SIDPlayer/discussions/2
